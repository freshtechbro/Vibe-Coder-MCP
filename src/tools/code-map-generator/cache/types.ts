/**
 * Type definitions for the file-based cache system.
 * This file contains interfaces for cache entries, metadata, and options.
 */

/**
 * Interface for a cache entry with metadata.
 */
export interface CacheEntry<T> {
  /**
   * The key for the cache entry.
   */
  key: string;

  /**
   * The value stored in the cache.
   */
  value: T;

  /**
   * The timestamp when the entry was created or last updated.
   */
  timestamp: number;

  /**
   * The expiration timestamp for the entry.
   */
  expiry: number;

  /**
   * Optional metadata for the entry.
   */
  metadata?: Record<string, unknown>;
}

/**
 * Interface for the cache metadata.
 */
export interface CacheMetadata {
  /**
   * The name of the cache.
   */
  name: string;

  /**
   * The number of entries in the cache.
   */
  size: number;

  /**
   * The timestamp when the cache was created.
   */
  createdAt: number;

  /**
   * The timestamp when the cache was last updated.
   */
  lastUpdated: number;

  /**
   * The keys of all entries in the cache.
   */
  keys: string[];

  /**
   * The maximum number of entries allowed in the cache.
   */
  maxEntries: number;

  /**
   * The maximum age of entries in milliseconds.
   */
  maxAge: number;

  /**
   * The total size of the cache in bytes.
   */
  sizeInBytes: number;
}

/**
 * Interface for cache configuration options.
 */
export interface CacheOptions {
  /**
   * The name of the cache.
   */
  name: string;

  /**
   * The directory where cache files are stored.
   */
  cacheDir: string;

  /**
   * The maximum number of entries allowed in the cache.
   */
  maxEntries?: number;

  /**
   * The maximum age of entries in milliseconds.
   */
  maxAge?: number;

  /**
   * Whether to validate cache entries on get.
   */
  validateOnGet?: boolean;

  /**
   * Whether to prune expired entries on startup.
   */
  pruneOnStartup?: boolean;

  /**
   * The interval in milliseconds for automatic pruning.
   */
  pruneInterval?: number;

  /**
   * A function to serialize cache values to strings.
   */
  serialize?: <T>(value: T) => string;

  /**
   * A function to deserialize strings to cache values.
   */
  deserialize?: <T>(serialized: string) => T;
}

/**
 * Interface for memory cache statistics.
 */
export interface MemoryCacheStats {
  /**
   * The number of hits (successful gets).
   */
  hits: number;

  /**
   * The number of entries in the cache.
   */
  size: number;

  /**
   * The total size of all entries in the cache.
   */
  totalSize: number;
}

/**
 * Interface for cache statistics.
 */
export interface CacheStats {
  /**
   * The name of the cache.
   */
  name: string;

  /**
   * The number of entries in the cache.
   */
  size: number;

  /**
   * The number of hits (successful gets).
   */
  hits: number;

  /**
   * The number of misses (unsuccessful gets).
   */
  misses: number;

  /**
   * The hit ratio (hits / (hits + misses)).
   */
  hitRatio: number;

  /**
   * The timestamp when the cache was created.
   */
  createdAt: number;

  /**
   * The timestamp when the cache was last updated.
   */
  lastUpdated: number;

  /**
   * The size of the cache in bytes.
   */
  sizeInBytes: number;

  /**
   * The total size of all entries in the cache.
   */
  totalSize?: number;

  /**
   * Memory cache statistics (if memory caching is enabled).
   */
  memoryStats?: MemoryCacheStats;
}

/**
 * Interface for memory statistics.
 */
export interface MemoryStats {
  /**
   * Formatted memory statistics for human readability.
   */
  formatted: {
    /**
     * Total system memory.
     */
    totalSystemMemory: string;

    /**
     * Free system memory.
     */
    freeSystemMemory: string;

    /**
     * Used system memory.
     */
    usedSystemMemory: string;

    /**
     * Memory usage percentage.
     */
    memoryUsagePercentage: string;

    /**
     * Memory status (normal, high, critical).
     */
    memoryStatus: 'normal' | 'high' | 'critical';

    /**
     * Process memory statistics.
     */
    process: {
      /**
       * Resident set size.
       */
      rss: string;

      /**
       * Total heap size.
       */
      heapTotal: string;

      /**
       * Used heap size.
       */
      heapUsed: string;

      /**
       * External memory.
       */
      external: string;

      /**
       * Array buffers memory.
       */
      arrayBuffers: string;
    };

    /**
     * V8 memory statistics.
     */
    v8: {
      /**
       * Heap size limit.
       */
      heapSizeLimit: string;

      /**
       * Total heap size.
       */
      totalHeapSize: string;

      /**
       * Used heap size.
       */
      usedHeapSize: string;

      /**
       * Heap size executable.
       */
      heapSizeExecutable: string;

      /**
       * Malloced memory.
       */
      mallocedMemory: string;

      /**
       * Peak malloced memory.
       */
      peakMallocedMemory: string;
    };

    /**
     * Cache statistics.
     */
    cache: {
      /**
       * Total cache size.
       */
      totalSize: string;

      /**
       * Number of caches.
       */
      cacheCount: number;
    };

    /**
     * Memory usage thresholds.
     */
    thresholds: {
      /**
       * High memory threshold.
       */
      highMemoryThreshold: string;

      /**
       * Critical memory threshold.
       */
      criticalMemoryThreshold: string;
    };
  };

  /**
   * Raw memory statistics.
   */
  raw: {
    /**
     * Total system memory in bytes.
     */
    totalSystemMemory: number;

    /**
     * Free system memory in bytes.
     */
    freeSystemMemory: number;

    /**
     * Memory usage percentage (0-1).
     */
    memoryUsagePercentage: number;

    /**
     * Process memory statistics.
     */
    processMemory: NodeJS.MemoryUsage;

    /**
     * V8 heap statistics.
     */
    heapStats: unknown;

    /**
     * V8 heap space statistics.
     */
    heapSpaceStats: unknown;
  };

  /**
   * Cache statistics.
   */
  cacheStats: CacheStats[];

  /**
   * Grammar statistics.
   */
  grammarStats: unknown;

  /**
   * Timestamp when the statistics were collected.
   */
  timestamp: number;
}
