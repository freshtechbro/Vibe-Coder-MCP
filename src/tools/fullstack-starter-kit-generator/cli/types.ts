export interface InitOptions {
  name: string;
  description?: string;
  template?: string;
  features?: string[];
  packages?: string[];
}

export interface PackageOptions {
  name: string;
  version: string;
  type: 'frontend' | 'backend' | 'library';
  path: string;
  dependencies?: Record<string, string>;
  devDependencies?: Record<string, string>;
}

export interface FrontendOptions {
  name: string;
  version: string;
  language: string;
  framework: {
    name: string;
    version: string;
  };
  styling?: {
    framework: string;
    version: string;
  };
}

export interface BackendOptions {
  name: string;
  version: string;
  language: string;
  framework: {
    name: string;
    version: string;
  };
  database?: {
    type: string;
    version: string;
  };
  orm?: {
    name: string;
    version: string;
  };
}
