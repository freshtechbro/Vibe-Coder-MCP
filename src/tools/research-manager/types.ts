import { z } from 'zod';

/**
 * Schema for research input validation
 */
export const researchInputSchema = z.object({
  query: z.string().min(5),
  depth: z.enum(['basic', 'detailed', 'comprehensive']).optional(),
  categories: z.array(z.string()).optional(),
});

/**
 * Type for validated research input
 */
export type ResearchInput = z.infer<typeof researchInputSchema>;

/**
 * Research result format
 */
export interface ResearchResult {
  summary: string;
  keyFindings: string[];
  detailedAnalysis: {
    sections: Array<{
      title: string;
      content: string;
    }>;
  };
  practicalApplications: string[];
  limitations: string[];
  recommendations: string[];
  metadata: {
    timestamp: string;
    query: string;
    depth?: string;
    categories?: string[];
  };
}

/**
 * Research analysis configuration
 */
export interface AnalysisConfig {
  depth: 'basic' | 'detailed' | 'comprehensive';
  format: 'markdown' | 'json' | 'yaml';
  maxTokens: number;
  temperature: number;
  categories?: string[];
}
