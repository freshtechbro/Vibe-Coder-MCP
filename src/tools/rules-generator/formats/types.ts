/**
 * Version information for a rule
 */
export interface RuleVersion {
  major: number;
  minor: number;
  patch: number;
  date: string;
  author?: string;
  changes: string[];
}

/**
 * Impact analysis for a rule
 */
export interface RuleImpact {
  scope: {
    files: string[];
    components: string[];
    services: string[];
  };
  effort: 'low' | 'medium' | 'high';
  risk: 'low' | 'medium' | 'high';
  benefits: string[];
  prerequisites: string[];
}

/**
 * Dependencies for a rule
 */
export interface RuleDependencies {
  requires: string[]; // IDs of rules that must be followed before this one
  conflicts: string[]; // IDs of rules that conflict with this one
  recommends: string[]; // IDs of rules that work well with this one
  alternatives: string[]; // IDs of alternative rules that could be used instead
}

/**
 * Change history entry for a rule
 */
export interface RuleChange {
  date: string;
  author?: string;
  type: 'added' | 'modified' | 'deprecated' | 'removed';
  description: string;
  version: string;
}

/**
 * Category relationship types
 */
export type CategoryRelationType =
  | 'parent'
  | 'child'
  | 'related'
  | 'requires'
  | 'conflicts';

/**
 * Category relationship
 */
export interface CategoryRelationship {
  type: CategoryRelationType;
  categoryId: string;
  description?: string;
}

/**
 * Category configuration
 */
export interface CategoryConfig {
  allowCustomRules: boolean;
  requireApproval: boolean;
  autoValidate: boolean;
  validateOnCommit: boolean;
  severity: 'error' | 'warning' | 'info';
}
