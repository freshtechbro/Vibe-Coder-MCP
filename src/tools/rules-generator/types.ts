import { z } from 'zod';

/**
 * Defines the structure of a rule template
 */
export interface RuleTemplate {
  name: string;
  description: string;
  sections: RuleSection[];
  variables: RuleVariable[];
  format: RuleFormat;
}

/**
 * Defines a section within a rule template
 */
export interface RuleSection {
  title: string;
  description: string;
  required: boolean;
  subsections?: RuleSection[];
}

/**
 * Defines a variable that can be used in templates
 */
export interface RuleVariable {
  name: string;
  description: string;
  type: 'string' | 'boolean' | 'number' | 'array';
  required: boolean;
  default?: unknown;
}

/**
 * Supported output formats
 */
export type RuleFormat = 'markdown' | 'yaml' | 'json' | 'html';

/**
 * Rule type categories
 */
export enum RuleCategory {
  Development = 'development',
  Documentation = 'documentation',
  Process = 'process',
  Custom = 'custom',
}

/**
 * Rule generation input schema
 */
export const ruleGenerationSchema = z.object({
  productDescription: z.string().min(10, {
    message: 'Product description must be at least 10 characters.',
  }),
  ruleTypes: z.array(z.string()).min(1, {
    message: 'At least one rule type must be specified.',
  }),
  format: z.enum(['markdown', 'yaml', 'json', 'html']).default('markdown'),
  customization: z.record(z.unknown()).optional(),
  context: z
    .object({
      repository: z.string().optional(),
      language: z.string().optional(),
      framework: z.string().optional(),
      team: z
        .object({
          size: z.number().optional(),
          experience: z.string().optional(),
        })
        .optional(),
    })
    .optional(),
});

export type RuleGenerationInput = z.infer<typeof ruleGenerationSchema>;

export interface RuleSemantics {
  intent: string;
  context: string;
  impact: string;
}

export interface RuleExample {
  description?: string;
  code?: string;
}

export interface Rule {
  name: string;
  semantics: {
    intent: string;
    context: string;
    impact: string;
  };
  rationale?: string;
  applicability?: string;
  filePatterns?: string[];
  examples?: Array<{
    description?: string;
    code?: string;
  }>;
}
