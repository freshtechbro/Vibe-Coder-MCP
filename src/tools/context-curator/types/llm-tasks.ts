/**
 * LLM task type definitions for the Context Curator tool.
 * This file contains enums, interfaces, and validation schemas for all LLM operations.
 */

import { z } from 'zod';
import { taskTypeSchema } from './context-curator.js';

/**
 * Enum for all Context Curator LLM tasks
 * Values must match exactly with llm_config.json task names
 */
export enum ContextCuratorLLMTask {
  INTENT_ANALYSIS = 'context_curator_intent_analysis',
  PROMPT_REFINEMENT = 'context_curator_prompt_refinement',
  FILE_DISCOVERY = 'context_curator_file_discovery',
  RELEVANCE_SCORING = 'context_curator_relevance_scoring',
  META_PROMPT_GENERATION = 'context_curator_meta_prompt_generation',
  ARCHITECTURAL_ANALYSIS = 'context_curator_architectural_analysis'
}

/**
 * Schema for intent analysis results
 */
export const intentAnalysisResultSchema = z.object({
  /** Detected task type */
  taskType: taskTypeSchema,
  /** Confidence in the analysis (0.0 to 1.0) */
  confidence: z.number().min(0).max(1),
  /** Reasoning for the analysis */
  reasoning: z.array(z.string()).min(1, 'At least one reasoning item is required'),
  /** Identified architectural components */
  architecturalComponents: z.array(z.string()),
  /** Scope assessment */
  scopeAssessment: z.object({
    complexity: z.enum(['simple', 'moderate', 'complex']),
    estimatedFiles: z.number().min(0),
    riskLevel: z.enum(['low', 'medium', 'high'])
  }),
  /** Suggested focus areas for analysis */
  suggestedFocusAreas: z.array(z.string()),
  /** Estimated effort level */
  estimatedEffort: z.enum(['low', 'medium', 'high', 'very_high'])
});

export type IntentAnalysisResult = z.infer<typeof intentAnalysisResultSchema> & {
  /** Enhanced project analysis data from Phase 2 */
  projectAnalysis?: ProjectTypeAnalysisResult;
  /** Enhanced language analysis data from Phase 2 */
  languageAnalysis?: LanguageAnalysisResult;
  /** Enhanced pattern analysis data from Phase 2 */
  patternAnalysis?: {
    patterns: string[];
    confidence: { [pattern: string]: number };
    evidence: { [pattern: string]: string[] };
  };
};

/**
 * Schema for file discovery results
 */
export const fileDiscoveryResultSchema = z.object({
  /** List of relevant files found */
  relevantFiles: z.array(z.object({
    path: z.string().min(1),
    priority: z.enum(['high', 'medium', 'low']),
    reasoning: z.string().min(1),
    confidence: z.number().min(0).max(1),
    estimatedTokens: z.number().min(0),
    modificationLikelihood: z.enum(['very_high', 'high', 'medium', 'low', 'very_low'])
  })),
  /** Total number of files analyzed */
  totalFilesAnalyzed: z.number().min(0),
  /** Processing time in milliseconds */
  processingTimeMs: z.number().min(0),
  /** Strategy used for file discovery */
  searchStrategy: z.enum(['semantic_similarity', 'keyword_matching', 'semantic_and_keyword', 'structural_analysis', 'multi_strategy']),
  /** Coverage metrics */
  coverageMetrics: z.object({
    totalTokens: z.number().min(0),
    averageConfidence: z.number().min(0).max(1)
  })
});

export type FileDiscoveryResult = z.infer<typeof fileDiscoveryResultSchema>;

/**
 * Type for individual file discovery result
 */
export type FileDiscoveryFile = {
  path: string;
  priority: 'high' | 'medium' | 'low';
  reasoning: string;
  confidence: number;
  estimatedTokens: number;
  modificationLikelihood: 'very_high' | 'high' | 'medium' | 'low' | 'very_low';
};

/**
 * Schema for prompt refinement results
 */
export const promptRefinementResultSchema = z.object({
  /** The refined prompt */
  refinedPrompt: z.string().min(1, 'Refined prompt cannot be empty'),
  /** Reasoning for enhancements */
  enhancementReasoning: z.array(z.string()),
  /** Context that was added */
  addedContext: z.array(z.string()),
  /** Original prompt length */
  originalLength: z.number().min(0),
  /** Refined prompt length */
  refinedLength: z.number().min(0),
  /** Improvement score (0.0 to 1.0) */
  improvementScore: z.number().min(0).max(1),
  /** Types of contextual enhancements made */
  contextualEnhancements: z.array(z.string())
});

export type PromptRefinementResult = z.infer<typeof promptRefinementResultSchema>;

/**
 * Schema for relevance scoring results
 */
export const relevanceScoringResultSchema = z.object({
  /** Individual file scores */
  fileScores: z.array(z.object({
    filePath: z.string().min(1),
    relevanceScore: z.number().min(0).max(1),
    confidence: z.number().min(0).max(1),
    reasoning: z.string().min(1),
    categories: z.array(z.string()).min(1, 'At least one category is required'),
    modificationLikelihood: z.enum(['very_high', 'high', 'medium', 'low', 'very_low']),
    estimatedTokens: z.number().min(0)
  })),
  /** Overall scoring metrics */
  overallMetrics: z.object({
    averageRelevance: z.number().min(0).max(1),
    totalFilesScored: z.number().min(0),
    highRelevanceCount: z.number().min(0),
    processingTimeMs: z.number().min(0)
  }),
  /** Strategy used for scoring */
  scoringStrategy: z.enum(['semantic_similarity', 'keyword_density', 'structural_importance', 'hybrid']),
  /** Optional chunking metadata for large file sets */
  chunkingUsed: z.boolean().optional(),
  /** Total number of chunks processed (only present when chunking is used) */
  totalChunks: z.number().min(1).optional(),
  /** Size of each chunk (only present when chunking is used) */
  chunkSize: z.number().min(1).optional()
});

export type RelevanceScoringResult = z.infer<typeof relevanceScoringResultSchema>;

/**
 * Schema for meta-prompt generation results
 */
export const metaPromptGenerationResultSchema = z.object({
  /** System prompt for AI agents */
  systemPrompt: z.string().min(1, 'System prompt cannot be empty'),
  /** User prompt for AI agents */
  userPrompt: z.string().min(1, 'User prompt cannot be empty'),
  /** Summary of the codebase context */
  contextSummary: z.string().min(1, 'Context summary cannot be empty'),
  /** Structured task decomposition */
  taskDecomposition: z.object({
    epics: z.array(z.object({
      id: z.string().min(1),
      title: z.string().min(1),
      description: z.string().min(1),
      estimatedComplexity: z.enum(['low', 'medium', 'high', 'very_high']),
      tasks: z.array(z.object({
        id: z.string().min(1),
        title: z.string().min(1),
        description: z.string().min(1),
        estimatedHours: z.number().min(0),
        dependencies: z.array(z.string()),
        subtasks: z.array(z.object({
          id: z.string().min(1),
          title: z.string().min(1),
          description: z.string().min(1),
          estimatedMinutes: z.number().min(0)
        }))
      }))
    }))
  }),
  /** Development guidelines */
  guidelines: z.array(z.string()),
  /** Estimated overall complexity */
  estimatedComplexity: z.enum(['low', 'medium', 'high', 'very_high']),
  /** Quality score of the generated meta-prompt */
  qualityScore: z.number().min(0).max(1),
  /** AI Agent response format specification */
  aiAgentResponseFormat: z.object({
    /** Description of the response format */
    description: z.string().min(1),
    /** Structured format template */
    format: z.string().min(1),
    /** Rules for AI agents to follow */
    rules: z.array(z.string().min(1))
  }).optional()
});

export type MetaPromptGenerationResult = z.infer<typeof metaPromptGenerationResultSchema>;

/**
 * Schema for architectural analysis results
 */
export const architecturalAnalysisResultSchema = z.object({
  /** Identified architectural patterns */
  architecturalPatterns: z.array(z.string()),
  /** Project dependencies */
  dependencies: z.array(z.object({
    name: z.string().min(1),
    version: z.string(),
    type: z.enum(['runtime', 'development', 'peer', 'optional']),
    importance: z.enum(['critical', 'important', 'optional', 'deprecated'])
  })),
  /** Code structure analysis */
  codeStructure: z.object({
    directories: z.array(z.string()),
    entryPoints: z.array(z.string()),
    configFiles: z.array(z.string())
  }),
  /** Architectural recommendations */
  recommendations: z.array(z.string()),
  /** Complexity assessment */
  complexityAssessment: z.object({
    overall: z.enum(['low', 'medium', 'high', 'very_high']),
    factors: z.array(z.string()),
    score: z.number().min(0).max(1)
  })
});

export type ArchitecturalAnalysisResult = z.infer<typeof architecturalAnalysisResultSchema>;

/**
 * Language analysis result with comprehensive language detection
 */
export interface LanguageAnalysisResult {
  /** Primary programming languages detected */
  languages: string[];
  /** File extensions found in the codebase */
  fileExtensions: string[];
  /** Grammar support status for each language */
  grammarSupport: { [language: string]: boolean };
  /** Distribution of languages by file count */
  languageDistribution: { [language: string]: number };
  /** Most prevalent language */
  primaryLanguage: string;
  /** Secondary languages (sorted by prevalence) */
  secondaryLanguages: string[];
  /** Framework indicators detected */
  frameworkIndicators: string[];
  /** Build system indicators detected */
  buildSystemIndicators: string[];
  /** Language confidence scores */
  languageConfidence: { [language: string]: number };
  /** Total files analyzed */
  totalFilesAnalyzed: number;
}

/**
 * Enhanced project type analysis result
 */
export interface ProjectTypeAnalysisResult {
  /** Primary project type */
  projectType: string;
  /** Secondary project types */
  secondaryTypes: string[];
  /** Confidence score for primary type */
  confidence: number;
  /** Evidence supporting the classification */
  evidence: string[];
  /** Framework stack detected */
  frameworkStack: string[];
  /** Architecture style indicators */
  architectureStyle: string[];
  /** Development environment indicators */
  developmentEnvironment: string[];
}

/**
 * Schema for LLM task errors
 */
export const llmTaskErrorSchema = z.object({
  /** The task that failed */
  task: z.nativeEnum(ContextCuratorLLMTask),
  /** Error message */
  message: z.string().min(1, 'Error message cannot be empty'),
  /** Error code */
  code: z.string().min(1),
  /** Additional error details */
  details: z.record(z.unknown()),
  /** Whether the error is recoverable */
  recoverable: z.boolean()
});

export type LLMTaskError = z.infer<typeof llmTaskErrorSchema>;

/**
 * Validation helper functions
 */
export const validateIntentAnalysisResult = (result: unknown): result is IntentAnalysisResult => {
  try {
    intentAnalysisResultSchema.parse(result);
    return true;
  } catch {
    return false;
  }
};

export const validateFileDiscoveryResult = (result: unknown): result is FileDiscoveryResult => {
  try {
    fileDiscoveryResultSchema.parse(result);
    return true;
  } catch {
    return false;
  }
};

export const validatePromptRefinementResult = (result: unknown): result is PromptRefinementResult => {
  try {
    promptRefinementResultSchema.parse(result);
    return true;
  } catch {
    return false;
  }
};

export const validateRelevanceScoringResult = (result: unknown): result is RelevanceScoringResult => {
  try {
    relevanceScoringResultSchema.parse(result);
    return true;
  } catch {
    return false;
  }
};

export const validateMetaPromptGenerationResult = (result: unknown): result is MetaPromptGenerationResult => {
  try {
    metaPromptGenerationResultSchema.parse(result);
    return true;
  } catch {
    return false;
  }
};

export const validateArchitecturalAnalysisResult = (result: unknown): result is ArchitecturalAnalysisResult => {
  try {
    architecturalAnalysisResultSchema.parse(result);
    return true;
  } catch {
    return false;
  }
};

/**
 * Helper function to get LLM task name from enum
 */
export const getLLMTaskName = (task: ContextCuratorLLMTask): string => {
  return task;
};

/**
 * Helper function to validate any LLM task result
 */
export const validateLLMTaskResult = (task: ContextCuratorLLMTask, result: unknown): boolean => {
  switch (task) {
    case ContextCuratorLLMTask.INTENT_ANALYSIS:
      return validateIntentAnalysisResult(result);
    case ContextCuratorLLMTask.FILE_DISCOVERY:
      return validateFileDiscoveryResult(result);
    case ContextCuratorLLMTask.PROMPT_REFINEMENT:
      return validatePromptRefinementResult(result);
    case ContextCuratorLLMTask.RELEVANCE_SCORING:
      return validateRelevanceScoringResult(result);
    case ContextCuratorLLMTask.META_PROMPT_GENERATION:
      return validateMetaPromptGenerationResult(result);
    case ContextCuratorLLMTask.ARCHITECTURAL_ANALYSIS:
      return validateArchitecturalAnalysisResult(result);
    default:
      return false;
  }
};
