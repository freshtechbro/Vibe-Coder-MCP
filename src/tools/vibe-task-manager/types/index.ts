// Re-export all types for easy importing
export * from './task.js';
export * from './dependency.js';
export * from './agent.js';
export * from './nl.js';
export * from './artifact-types.js';
export * from './project-context.js';
