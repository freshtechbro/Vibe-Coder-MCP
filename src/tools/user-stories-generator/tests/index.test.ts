import { McpError } from '@modelcontextprotocol/sdk/types.js';
import { describe, it, expect, vi, beforeEach } from 'vitest';

import { mockOpenRouterConfig } from '../../../test-utils/mock-configs.js';
import { OpenRouterConfig } from '../../../types/workflow.js';
import * as llmHelper from '../../../utils/llmHelper.js';
import * as researchHelper from '../../../utils/researchHelper.js';
import { generateUserStories } from '../index.js';

// Mock dependencies
vi.mock('../../../utils/researchHelper.js');
vi.mock('../../../utils/llmHelper.js');
vi.mock('../../../logger.js');

describe('User Stories Generator Tool', () => {
  // Use standardized mock configuration
  const mockConfig = mockOpenRouterConfig;

  // Mock execution context
  const mockContext = {
    sessionId: 'test-session-id',
    userId: 'test-user-id',
    jobId: 'test-job-id',
  };

  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('should generate user stories successfully', async () => {
    const mockResearchResult = 'Mock research data about user needs';
    const mockUserStories =
      '# User Stories\n1. As a user, I want to...\n2. As a user, I want to...';
    vi.mocked(researchHelper.performResearchQuery).mockResolvedValue(
      mockResearchResult
    );
    vi.mocked(llmHelper.performDirectLlmCall).mockResolvedValue(
      mockUserStories
    );

    const result = await generateUserStories(
      { productDescription: 'Test Product' },
      mockConfig,
      mockContext
    );

    expect(result.isError).toBe(false);
    expect(result.content[0]?.text).toBe(mockUserStories);
    expect(researchHelper.performResearchQuery).toHaveBeenCalledWith(
      expect.stringContaining('Test Product'),
      mockConfig,
      mockContext
    );
    expect(llmHelper.performDirectLlmCall).toHaveBeenCalledWith(
      expect.stringContaining(mockResearchResult),
      expect.stringContaining(
        'As a [user type], I want to [action] so that [benefit]'
      ),
      mockConfig,
      'user_stories_generation',
      0.3,
      mockContext
    );
  });

  it('should respect maxStories parameter', async () => {
    const mockResearchResult = 'Mock research data';
    const mockUserStories =
      '# User Stories\n1. As a user, I want to...\n2. As a user, I want to...';
    vi.mocked(researchHelper.performResearchQuery).mockResolvedValue(
      mockResearchResult
    );
    vi.mocked(llmHelper.performDirectLlmCall).mockResolvedValue(
      mockUserStories
    );

    const maxStories = 5;
    await generateUserStories(
      { productDescription: 'Test Product', maxStories },
      mockConfig,
      mockContext
    );

    expect(llmHelper.performDirectLlmCall).toHaveBeenCalledWith(
      expect.stringContaining(`up to ${maxStories} stories`),
      expect.stringContaining(
        'As a [user type], I want to [action] so that [benefit]'
      ),
      mockConfig,
      'user_stories_generation',
      0.3,
      mockContext
    );
  });

  it('should handle empty research results', async () => {
    // Use empty string instead of null to satisfy type constraints
    vi.mocked(researchHelper.performResearchQuery).mockResolvedValue('');

    const result = await generateUserStories(
      { productDescription: 'Test Product' },
      mockConfig,
      mockContext
    );

    expect(result.isError).toBe(true);
    expect(result.content[0]?.text).toContain(
      'Research query returned no results'
    );
    const errorDetails = result.errorDetails as McpError;
    expect(errorDetails?.message).toContain(
      'Research query returned no results'
    );
  });

  it('should handle LLM generation failure', async () => {
    const mockResearchResult = 'Mock research data';
    vi.mocked(researchHelper.performResearchQuery).mockResolvedValue(
      mockResearchResult
    );
    // Use empty string instead of null to satisfy type constraints
    vi.mocked(llmHelper.performDirectLlmCall).mockResolvedValue('');

    const result = await generateUserStories(
      { productDescription: 'Test Product' },
      mockConfig,
      mockContext
    );

    expect(result.isError).toBe(true);
    expect(result.content[0]?.text).toContain('User stories generation failed');
    const errorDetails = result.errorDetails as McpError;
    expect(errorDetails?.message).toContain('User stories generation failed');
  });

  it('should handle research query errors', async () => {
    vi.mocked(researchHelper.performResearchQuery).mockRejectedValue(
      new Error('Research failed')
    );

    const result = await generateUserStories(
      { productDescription: 'Test Product' },
      mockConfig,
      mockContext
    );

    expect(result.isError).toBe(true);
    expect(result.content[0]?.text).toContain('Research failed');
    const errorDetails = result.errorDetails as McpError;
    expect(errorDetails?.message).toContain('Research failed');
  });

  it('should handle LLM call errors', async () => {
    const mockResearchResult = 'Mock research data';
    vi.mocked(researchHelper.performResearchQuery).mockResolvedValue(
      mockResearchResult
    );
    vi.mocked(llmHelper.performDirectLlmCall).mockRejectedValue(
      new Error('LLM failed')
    );

    const result = await generateUserStories(
      { productDescription: 'Test Product' },
      mockConfig,
      mockContext
    );

    expect(result.isError).toBe(true);
    expect(result.content[0]?.text).toContain('LLM failed');
    const errorDetails = result.errorDetails as McpError;
    expect(errorDetails?.message).toContain('LLM failed');
  });

  it('returns async job message with wait instruction', async () => {
    const params = { productDescription: 'Test Product', async: true };
    const result = await generateUserStories(params, {}, mockContext);
    expect(result.isError).toBe(false);
    expect(result.content[0].text).toContain('Your request has been received and is being processed as an async job.');
    expect(result.content[0].text).toContain('Job ID:');
    expect(result.content[0].text).toContain('Please wait a moment for the task to complete before attempting to retrieve the job result.');
    expect(result.content[0].text).toContain('To check the status or result of this job, send the following prompt:');
    expect(result.content[0].text).toContain('user-stories-job-result');
  });
});
