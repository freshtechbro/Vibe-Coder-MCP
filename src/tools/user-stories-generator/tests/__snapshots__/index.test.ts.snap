// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`User Stories Generator (Async) > should set final job result content matching snapshot (async) > User Stories Generator Content 1`] = `
"# User Stories: Event Ticketing App

## Epic: User Registration
### User Story: Account Creation..."
`;

exports[`User Stories Generator > should generate user stories content matching snapshot > User Stories Generator Content 1`] = `
"# User Stories: A ticketing application for event management...

# User Stories: Event Ticketing App

## Epic: User Registration
### User Story: Account Creation
**ID:** US-101
**Title:** Create User Account

**Story:**
> As a **new event attendee**,
> I want to **create a personal account**
> So that **I can purchase tickets and track my orders**.

**Acceptance Criteria:**
* GIVEN I am on the registration page WHEN I enter valid details THEN my account is created successfully.
* GIVEN I am registering WHEN I enter a password THEN it is validated for minimum security requirements.

**Priority:** High
**Dependencies:** None"
`;
